src/Microsoft.NetFramework.Analyzers/Core/DoNotCatchCorruptedStateExceptions.cs

```diff
@@ -33,108 +31,24 @@ public sealed class DoNotCatchCorruptedStateExceptionsAnalyzer : DiagnosticAnaly
-                                method.ToDisplayString()));
-                        }
-                    }
-                });
-            });
         }
 
-        private static bool ContainsHandleProcessCorruptedStateExceptionsAttribute(IMethodSymbol method, CompilationSecurityTypes compilationTypes)
+        protected override bool ShouldCheckCompilationUnit(Compilation compilation)
         {
-            ImmutableArray<AttributeData> attributes = method.GetAttributes();
-            return attributes.Any(
-                attribute => attribute.AttributeClass.Equals(compilationTypes.HandleProcessCorruptedStateExceptionsAttribute));
+            return SecurityTypes.HandleProcessCorruptedStateExceptionsAttribute(compilation) != null;
         }
 
-        /// <summary>
-        /// Walks an IOperation tree to find catch-all blocks that contain no "throw;" statements.
-        /// </summary>
-        private class EmptyThrowInsideCatchAllWalker : OperationWalker
+        protected override bool ShouldCheckMethod(Compilation compilation, IMethodSymbol method)
         {
-            private readonly CompilationSecurityTypes _compilationTypes;
-            private readonly Stack<bool> _seenEmptyThrowInCatchClauses = new Stack<bool>();
-
-            public ISet<ICatchClauseOperation> CatchAllCatchClausesWithoutEmptyThrow { get; } = new HashSet<ICatchClauseOperation>();
-
-            public EmptyThrowInsideCatchAllWalker(CompilationSecurityTypes compilationTypes)
-            {
-                _compilationTypes = compilationTypes;
-            }
-
-            public override void VisitAnonymousFunction(IAnonymousFunctionOperation operation)
-            {
-                // for now there doesn't seem to be any way to annotate lambdas with attributes
-            }
-
-            public override void VisitCatchClause(ICatchClauseOperation operation)
-            {
-                _seenEmptyThrowInCatchClauses.Push(false);
-
-                Visit(operation.Filter);
-                Visit(operation.Handler);
-
-                bool seenEmptyThrow = _seenEmptyThrowInCatchClauses.Pop();
-
-                if (IsCaughtTypeTooGeneral(operation.ExceptionType) && !seenEmptyThrow)
-                {
-                    CatchAllCatchClausesWithoutEmptyThrow.Add(operation);
-                }
-            }
-
-            public override void VisitThrow(IThrowOperation operation)
-            {
-                if (operation.Exception == null && _seenEmptyThrowInCatchClauses.Count > 0 && !_seenEmptyThrowInCatchClauses.Peek())
-                {
-                    _seenEmptyThrowInCatchClauses.Pop();
-                    _seenEmptyThrowInCatchClauses.Push(true);
-                }
-
-                base.VisitThrow(operation);
-            }
+            var enablingAttribute = SecurityTypes.HandleProcessCorruptedStateExceptionsAttribute(compilation);
```

Comment:
Can we follow existing pattern used by analyzers and fetch this well known type in the CompilationStartAction, and pass `enablingAttribute` type as a parameter to `ShouldCheckMethod`?

Reply:
I wasn't sure whether the extra calls to GetTypeByMetadataName would be a significant performance hit. Your suggestion favors performance at a small design cost, but it sounds like it's the better option.

Reply:
We have had multiple discussions on this in the past and have decided to take the caching approach in compilation start action - I believe @sharwell has shown the performance can degrade in presence of multiple analyzers taking your approach as `GetTypeByMetadataName` has a bounded cache.

Reply:
I've removed the virtual `ShouldCheckCompilationUnit` and `ShouldCheckMethod` methods and moved the "only check methods with a certain attribute" logic into the shared base class.